include/deal.II/matrix_free/matrix_free.h

```diff
@@ -2919,130 +2919,155 @@ namespace internal
+                                    IsBlockVector<VectorStruct>::value>(),
+                                    exchanger);
   }
 
 
 
   template <int dim, typename Number, typename Number2>
   inline
-  void reset_ghost_values (const LinearAlgebra::distributed::Vector<Number> &vec,
-                           VectorDataExchange<dim,Number2> &exchanger)
+  void update_ghost_values_start (const LinearAlgebra::distributed::Vector<Number> &vec,
+                                  VectorDataExchange<dim,Number2> &exchanger,
+                                  const unsigned int channel = 0)
   {
-    exchanger.reset_ghost_values(vec);
+    exchanger.update_ghost_values_start(channel, vec);
   }
 
 
 
   template <int dim, typename VectorStruct, typename Number>
   inline
-  void reset_ghost_values (const std::vector<VectorStruct> &vec,
-                           VectorDataExchange<dim,Number> &exchanger)
+  void update_ghost_values_start (const std::vector<VectorStruct> &vec,
+                                  VectorDataExchange<dim,Number> &exchanger)
   {
+    unsigned int component_index = 0;
     for (unsigned int comp=0; comp<vec.size(); comp++)
-      reset_ghost_values(vec[comp], exchanger);
+      {
+        update_ghost_values_start(vec[comp], exchanger, component_index);
+        component_index += n_components(vec[comp]);
+      }
   }
 
 
 
   template <int dim, typename VectorStruct, typename Number>
   inline
-  void reset_ghost_values (const std::vector<VectorStruct *> &vec,
-                           VectorDataExchange<dim,Number> &exchanger)
+  void update_ghost_values_start (const std::vector<VectorStruct *> &vec,
+                                  VectorDataExchange<dim,Number> &exchanger)
   {
+    unsigned int component_index = 0;
     for (unsigned int comp=0; comp<vec.size(); comp++)
-      reset_ghost_values(*vec[comp], exchanger);
+      {
+        update_ghost_values_start(*vec[comp], exchanger, component_index);
+        component_index += n_components(*vec[comp]);
+      }
   }
 
 
 
   template <int dim, typename VectorStruct, typename Number>
   inline
-  void reset_ghost_values_block (const VectorStruct &vec,
-                                 std::integral_constant<bool,true>,
-                                 VectorDataExchange<dim,Number> &exchanger)
+  void update_ghost_values_start_block (const VectorStruct &vec,
+                                        const unsigned int channel,
+                                        std::integral_constant<bool,true>,
+                                        VectorDataExchange<dim,Number> &exchanger)
   {
-    for (unsigned int i=0; i<vec.n_blocks(); ++i)
-      reset_ghost_values(vec.block(i), exchanger);
+    if (get_communication_block_size(vec) < vec.n_blocks())
+      vec.update_ghost_values();
+    else
+      for (unsigned int i=0; i<vec.n_blocks(); ++i)
+        update_ghost_values_start(vec.block(i), exchanger, channel+i);
   }
```

Comment:
Everything above (apart from `get_communication_block_size`) is only reordering, isn't it?

Reply:
yes